### ISSUE TITLE
java.lang.NoClassDefFoundError: za/co/absa/cobrix/cobol/parser/encoding/Encoding

### ISSUE THREAD

**Issue by ryantnedbank** [AbsaOSS/cobrix#1]
I have done an mvn package to build your code.
I referenced the three produced jars in spark-shell:
spark-cobol-0.2.6-SNAPSHOT.jar
spark-cobol-0.2.6-SNAPSHOT-javadoc.jar
spark-cobol-0.2.6-SNAPSHOT-sources.jar

Then I invoke spark-shell:
spark-shell --jars /path_to_cobrix_jars/*.jar

I am getting an error that a particular class is missing.
Can you guide me?

scala> :load /home/private/Documents/SparkSQLExample_cobol.scala
Loading /home/private/Documents/SparkSQLExample_cobol.scala...
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._
import org.apache.spark.sql.{Row, DataFrame, SQLContext}
import org.apache.spark.sql.SQLContext
import za.co.absa.cobrix._
spark: org.apache.spark.sql.SparkSession.type = org.apache.spark.sql.SparkSession$@37f41a81
res0: spark.Builder = org.apache.spark.sql.SparkSession$Builder@5ee581db
res1: spark.Builder = org.apache.spark.sql.SparkSession$Builder@5ee581db
2018-09-11 10:16:51 WARN  SparkSession$Builder:66 - Using an existing SparkSession; some configuration may not take effect.
res2: org.apache.spark.sql.SparkSession = org.apache.spark.sql.SparkSession@5a8656a2
warning: there was one deprecation warning; re-run with -deprecation for details
sqlContext: org.apache.spark.sql.SQLContext = org.apache.spark.sql.SQLContext@4070c4ff
java.lang.NoClassDefFoundError: za/co/absa/cobrix/cobol/parser/encoding/Encoding
  at za.co.absa.cobrix.spark.cobol.source.DefaultSource.createFixedLengthReader(DefaultSource.scala:82)
  at za.co.absa.cobrix.spark.cobol.source.DefaultSource.buildEitherReader(DefaultSource.scala:69)
  at za.co.absa.cobrix.spark.cobol.source.DefaultSource.createRelation(DefaultSource.scala:50)
  at za.co.absa.cobrix.spark.cobol.source.DefaultSource.createRelation(DefaultSource.scala:45)
  at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:340)
  at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:239)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:227)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:174)
  ... 77 elided
Caused by: java.lang.ClassNotFoundException: za.co.absa.cobrix.cobol.parser.encoding.Encoding
  at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  ... 85 more

Code executed is below:
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._
import org.apache.spark.sql.{Row, DataFrame, SQLContext}
import org.apache.spark.sql.SQLContext
import za.co.absa.cobrix._

val spark = SparkSession
.builder()
.appName("Spark SQL basic example")
.getOrCreate()

val sqlContext = new org.apache.spark.sql.SQLContext(sc)

val DataFrame = sqlContext.read.format("cobol").option("copybook", "/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.cpy").load("/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.dat")

**Comment by yruslan** [comment]
`za.co.absa.cobrix.cobol.parser.encoding.Encoding` is part of COBOL parser.
It seems you need to package cobol-parser module as well and put it to the jars folder.
The parser is in a separate module so people can use use it independently of Spark.

**Comment by yruslan** [comment]
Please let me know if it is sorted or more detailed guide is needed.

**Comment by yruslan** [comment]
I see. It seems like the short name for the data source haven't picked up by spark-shell. Please, try this code instead:

 DataFrame = **spark**
    .read
    .format("**za.co.absa.cobrix.spark.cobol.source**")
    .option("copybook", "/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.cpy")
    .load("/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.dat")

Note the usage of the spark session instead of sqlContext and the full name of the data source.

Usually we create a project having a Maven dependency on Cobrix and use Maven's Shade plugin to create an uber jar for submitting Spark jobs on a cluster. 
An example of such a Spark Job is here (without Cobrix dependency): https://github.com/yruslan/spark-skel
The job could also be ran locally from an IDE by running SampleSparkAppRunner object as a unit test. This will bring up all dependencies that have 'provided' scope and are not included in uber jar. 

We will try your scenario and update the README.md accordingly.

**Comment by ryantnedbank** [comment]
Hi Ruslan

I have tried this.  It doesn't work.  Please guide me.
spark-shell --jars /home/private/Documents/AbsaOSS_cobrix/cobrix-master/spark-cobol/target/*.jar

2018-09-14 11:17:23 WARN  Utils:66 - Your hostname, localhost.localdomain resolves to a loopback address: 127.0.0.1; using 10.0.2.15 instead (on interface enp0s3)
2018-09-14 11:17:23 WARN  Utils:66 - Set SPARK_LOCAL_IP if you need to bind to another address
2018-09-14 11:17:24 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context Web UI available at http://10.0.2.15:4040
Spark context available as 'sc' (master = local[*], app id = local-1536916658103).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 2.3.1
      /_/
         
Using Scala version 2.11.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_181)
Type in expressions to have them evaluated.
Type :help for more information.

scala> :load /home/private/Documents/SparkSQLExample_cobol.scala

Loading /home/private/Documents/SparkSQLExample_cobol.scala...
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._
import org.apache.spark.sql.{Row, DataFrame, SQLContext}
import org.apache.spark.sql.SQLContext
import za.co.absa.cobrix._
spark: org.apache.spark.sql.SparkSession.type = org.apache.spark.sql.SparkSession$@3593e074
res0: spark.Builder = org.apache.spark.sql.SparkSession$Builder@78bd02c8
res1: spark.Builder = org.apache.spark.sql.SparkSession$Builder@78bd02c8
2018-09-14 11:18:08 WARN  SparkSession$Builder:66 - Using an existing SparkSession; some configuration may not take effect.
res2: org.apache.spark.sql.SparkSession = org.apache.spark.sql.SparkSession@27055a2a
warning: there was one deprecation warning; re-run with -deprecation for details
sqlContext: org.apache.spark.sql.SQLContext = org.apache.spark.sql.SQLContext@727f36dd
DataFrame: org.apache.spark.sql.SparkSession.type = org.apache.spark.sql.SparkSession$@3593e074
<console>:36: error: value read is not a member of object org.apache.spark.sql.SparkSession
       DataFrame.read
                 ^
<console>:36: error: value format is not a member of object org.apache.spark.sql.SparkSession
       DataFrame.format("za.co.absa.cobrix.spark.cobol.source")
                 ^
<console>:36: error: value option is not a member of object org.apache.spark.sql.SparkSession
       DataFrame.option("copybook", "/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.cpy")
                 ^
<console>:36: error: value load is not a member of object org.apache.spark.sql.SparkSession
       DataFrame.load("/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.dat")

The contents of SparkSQLExample_cobol.scala
---------------------------------------------------
// A package is an organization unit that can contain entities such as classes, objects and other packages
// Multiple nested packages can be declared by inserting "." between each package
// package org.apache.spark.examples.sql       You can't define a package when running interactively in spark-shell
// SparkSession class is the entry point to all functionality in Spark
import org.apache.spark.sql.SparkSession
//Row companion object offers factory methods to create Row instances from a collection of elements ( apply ), a sequence of elements ( fromSeq ) and tuples ( fromTuple )
import org.apache.spark.sql.types._
import org.apache.spark.sql.{Row, DataFrame, SQLContext}
import org.apache.spark.sql.SQLContext
import za.co.absa.cobrix._
//You must invoke spark-shell with extra jar files --> spark-shell --jars /home/private/Documents/AbsaOSS_cobrix/cobrix-master/spark-cobol/target/*.jar

val spark = SparkSession
.builder()
.appName("Spark SQL basic example")
.getOrCreate()

val sqlContext = new org.apache.spark.sql.SQLContext(sc)

val DataFrame = spark
.read
.format("za.co.absa.cobrix.spark.cobol.source")
.option("copybook", "/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.cpy")
.load("/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.dat")

//val DataFrame = sqlContext.read.format("cobol").option("copybook", "/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.cpy").load("/home/private/Documents/Data_Files/Cobol_EBCDIC/pnic00.wspack.sam010.d0.sq2700.D180630.dat")

**Comment by yruslan** [comment]
Here is another example. It uses 'collapse_root' feature to expand root struct field:
```
scala> val df = spark.read.format("cobol").option("copybook", "/user/yruslan/example1/test3_copybook.cob").option("schema_retention_policy", "collapse_root").load("/user/yruslan/example1/data")
df: org.apache.spark.sql.DataFrame = [CURRENCY: string, SIGNATURE: string ... 4 more fields]

scala> df.show(false)
+--------+---------+-------------+-----------+----------+----------+
|CURRENCY|SIGNATURE|COMPANY_NAME |COMPIANY_ID|WEALTH_QFY|AMOUNT    |
+--------+---------+-------------+-----------+----------+----------+
|GBP     |S9276511 |Beierbauh.   |0123330087 |1         |89341.00  |
|ZAR     |S9276511 |Envision Inc.|0039003991 |1         |2634633.00|
|USD     |S9276511 |Beierbauh.   |0038903321 |0         |75.71     |
|ZAR     |S9276511 |Beierbauh.   |0123330087 |0         |215.39    |
|ZAR     |S9276511 |Pear GMBH.   |0002377771 |0         |716.29    |
|CHF     |S9276511 |Envision Inc.|0039003991 |1         |983.99    |
|ZAR     |S9276511 |ABCD Ltd.    |0039887123 |1         |2527640.00|
|ZAR     |S9276511 |Envision Inc.|0039003991 |1         |5434295.00|
|ZAR     |S9276511 |Pear GMBH.   |0002377771 |0         |520.80    |
|ZAR     |S9276511 |Envision Inc.|0039003991 |0         |154.48    |
|CHF     |S9276511 |ABCD Ltd.    |0039887123 |0         |790.70    |
|EUR     |S9276511 |Pear GMBH.   |0002377771 |1         |679.64    |
|ZAR     |S9276511 |ABCD Ltd.    |0039887123 |1         |721.05    |
|ZAR     |S9276511 |Pear GMBH.   |0002377771 |0         |646.33    |
|ZAR     |S9276511 |Pear GMBH.   |0002377771 |1         |477.14    |
|ZAR     |S9276511 |Prime Bank   |0092317899 |1         |643.94    |
|ZAR     |S9276511 |Xingzhoug    |8822278911 |1         |998.03    |
|USD     |S9276511 |Beierbauh.   |0123330087 |1         |848.88    |
|USD     |S9276511 |Beierbauh.   |0123330087 |0         |664.11    |
|ZAR     |S9276511 |Beierbauh.   |0123330087 |1         |55262.00  |
+--------+---------+-------------+-----------+----------+----------+
only showing top 20 rows
``` 

Btw, this is an autogenerated ebcidic file. The generator will be included in the next release.
